ambry-account/src/main/java/com/github/ambry/account/BackupFileManager.java

```diff
@@ -0,0 +1,459 @@
+    this.accountServiceMetrics = accountServiceMetrics;
+    this.config = config;
+    backupDirPath = config.backupDir.isEmpty() ? null : Files.createDirectories(Paths.get(config.backupDir));
+    backupFileInfos = new ConcurrentSkipListMap<>();
+    cleanupBackupFiles();
+  }
+
+  /**
+   * Clean up the backup files. It cleans up
+   * <ul>
+   *   <li>Temporary files when creating a backup file</li>
+   *   <li>Old backup files with version number</li>
+   *   <li>Backup files in old format without version number</li>
+   * </ul>
+   */
+  private void cleanupBackupFiles() {
+    if (backupDirPath == null) {
+      return;
+    }
+    File backupDir = backupDirPath.toFile();
+
+    // First get all the file with temp file suffix and remove all of them
+    FileFilter tempFileFilter = (File pathname) -> pathname.getName().endsWith(SEP + TEMP_FILE_SUFFIX);
+    for (File file : backupDir.listFiles(tempFileFilter)) {
+      logger.trace("Delete temp file " + file.getName());
+      tryDeleteFile(file);
+    }
+
+    // Then get all the file with version number and local timestamp
+    FileFilter versionFileFilter = (File pathname) -> versionFilenamePattern.matcher(pathname.getName()).find();
+    for (File file : backupDir.listFiles(versionFileFilter)) {
+      Matcher m = versionFilenamePattern.matcher(file.getName());
+      m.find();
+      logger.trace("Starting processing version backup file " + file.getName());
+      int version = Integer.parseInt(m.group(1));
+      long modifiedTimeInSecond = LocalDateTime.parse(m.group(2), TIMESTAMP_FORMATTER).toEpochSecond(zoneOffset);
+      BackupFileInfo currentBackup = new BackupFileInfo(version, file.getName(), modifiedTimeInSecond);
+
+      if (backupFileInfos.size() < config.maxBackupFileCount) {
+        // When the number of backup files are under the maximum value, just add the current backupFile in the map.
+        backupFileInfos.put(version, currentBackup);
+      } else {
+        // When the number of backup files exceeds the maximum value, we have to remove the one backupFile.
+        if (backupFileInfos.firstEntry().getKey() < version) {
+          // The current backupFile's version is larger than the smallest one in the backupFileInfos map, then remove the
+          // smallest backupFile and it's entry from map and add the current backupFile in the map.
+          Map.Entry<Integer, BackupFileInfo> entry = backupFileInfos.firstEntry();
+          BackupFileInfo toRemove = entry.getValue();
+          logger.trace("Remove the oldest backup {} at version {}", toRemove.getFilename(), toRemove.getVersion());
+          tryRemoveBackupFile(entry.getValue());
+          backupFileInfos.remove(entry.getKey());
+          backupFileInfos.put(version, currentBackup);
+        } else {
+          // The current backupFile's version is smaller than the smallest version in the map, then remove the
+          // current backupFile.
+          tryRemoveBackupFile(currentBackup);
```

Comment:
The listFiles function doesn't sort the result based on the filename, it just returns all the files. Imagine we used to have the maximum number of backup file as 100, then somehow we decide to decrease the number to 50, that means, while ambry-frontend starts up, we might have to remove 50 of them. And since the files are not ordered, the version would be smaller than the smallest number in the map.